keps/sig-storage/20190120-ExecutionHook-apichange.md

```diff
@@ -0,0 +1,723 @@
+        // +optional
+        Policy *HookPolicy
+
+        // This is required.
+        PreAction *ExecutionHookAction
+
+        // +optional
+        PostAction *ExecutionHookAction
+
+        // Time between PreAction (determined by the first ActionTimeStamp for PreAction) and
+        // PostAction (determined by the last ActionTimeStamp for PostAction).
+        // ExecutionHook controller will check this timeout. If API server dies, hook controller
+        // will still trigger PostAction command if this field is set, timeout happens, and
+        // PreAction is called (determined by the first ActionTimestamp for PreAction)
+        // +optional
+        ExpirationSeconds *int64
+}
+```
+
+Here is the definition for HookPolicy:
+```
+type HookPolicy string
+
+// These are valid policies of ExecutionHook
+const (
+        // Execute commands only once on any one selected pod
+        HookExecuteOnce HookPolicy = "ExecuteOnce"
+        // Execute commands on all selected pods
+        HookExecuteAll HookPolicy = "ExecuteAll"
+)
+```
+
+The following explains the details of how to use ExecutionHook for volume snapshotting use case.
+
+API Change: Add ExecutionHookInfos to VolumeSnapshotSpec:
+```
+// VolumeSnapshotSpec describes the common attributes of a volume snapshot
+type VolumeSnapshotSpec struct {
+        // Source has the information about where the snapshot is created from.
+        // In Alpha version, only PersistentVolumeClaim is supported as the source.
+        // If not specified, user can create VolumeSnapshotContent and bind it with VolumeSnapshot
+        //  manually.
+        // +optional
+        Source *core_v1.TypedLocalObjectReference `json:"source" protobuf:"bytes,1,opt,name=source"`
+
+        // SnapshotContentName binds the VolumeSnapshot object with the VolumeSnapshotContent
+        // +optional
+        SnapshotContentName string `json:"snapshotContentName" protobuf:"bytes,2,opt,name=snapshotContentName"`
+
+        // Name of the VolumeSnapshotClass used by the VolumeSnapshot. If not specified, a default
+        // snapshot class will be used if it is available.
+        // +optional
+        VolumeSnapshotClassName *string `json:"snapshotClassName" protobuf:"bytes,3,opt,name=snapshotClassName"`
+
+        // ExecutionHookInfos is a new field in VolumeSnapshotSpec
+        // List of ExecutionHookInfo. It is used to build ExecutionHooks
+	// Hooks will be run sequencially now.
+	// TODO: In the future, we may add ExecutionOrdering to indicate the ordering
+	// of running the hooks.
+        // +optional
+        ExecutionHookInfos []ExecutionHookInfo
+}
+```
+
+ExectionHookNames will be copied to VolumeSnapshotStatus when the volume snapshot
+controller creates the hooks. If a hook execution times out on a container, execution
+hook controller won't delete the hook. User can look up the hook status and clean up
```

Comment:
If the hook is not deleted and `ExpirationSeconds` is not set, the `PostAction` will never be triggered? This is tricky as the `PreAction` might have been executed, in which case the `PostAction` (if any) should be executed even if we don't know for sure if the execution of the `PreAction` was successful or not.

Reply:
but without ExpirationSeconds set, controller cannot decide when to trigger PostAction.

Reply:
Should we have a default ExpirationSeconds set?  It will be hard to find an optimal default though as every application is different.